lock/lock.go

```diff
@@ -0,0 +1,333 @@
+package lock
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/gruntwork-io/go-commons/retry"
+	"github.com/sirupsen/logrus"
+	"time"
+)
+
+// Terraform requires the DynamoDB table to have a primary key with this name
+const attributeLockId = "LockID"
+
+// Default is to retry for up to 5 minutes
+const maxRetriesWaitingForTableToBeActive = 30
+const sleepBetweenTableStatusChecks = 10 * time.Second
+
+const dynamoDbPayPerRequestBillingMode = "PAY_PER_REQUEST"
+
+type Options struct {
+	Logger *logrus.Logger
+	LockString string
+	LockTable string
+	AwsRegion string
+	MaxRetries int
+	SleepBetweenRetries time.Duration
+	CreateTableIfMissing bool
```

Comment:
NIT: add `//` comments next to each field explaining what it is.

NIT: put these in a more "story like" order. E.g., `Logger` is least important, so put it at the end. `LockTable` and `AwsRegion` should prob be at the top; then `LockString`; then the two retry params; etc.

Reply:
I added the comments - I'll have to re-read them once again though to see if they make sense